### PR TITLE
#zqZkd9M8 Roles and Permissions

### PR DESCRIPTION
**Pull request by ncutixavier**
### **What about this PR?**
This PR is about to set permission based on role

### **Description of Task to be completed**
- Set permission based on role
- Appropriate message to the user

### **Quick Test with documentation**
- Use the [documentation](https://phantom-caba-ft-roles-a-hhgudv.herokuapp.com/api-docs/) link
- Login with the credential below
- Copy the token and use it to update user. Token should be like this `Bearer <token>`

### **How should be manually tested?**
- Clone the repo
- checkout the branch `ft-roles-and-permissions-zqZkd9M8`
- Install all packages
- Set up dotenv with `.env.example` as reference
- Run this command `npm run migrateDB` to make migration and seeding
- To access this endpoint **`GET /api/v1/auth/updateuser`** you have to login with these credentials: 
{
    "email":"admin@phantom.com",
    "password":"admin"
}
- Copy token and paste it in documentation or in header if you are using postman. Hint: `Bearer <token>`
- Try update user with this endpoint **`GET /api/v1/auth/updateuser`**

### **Relevant pivots tracker**
[#zqZkd9M8](https://trello.com/c/zqZkd9M8/9-create-roles-and-set-permissions)

### CODE REVIEW

File: src/routes/index.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+router.use('/api/v1/auth', userRoute)
+
+router.all('*', (req, res) => {
+    res.status(404).json({
+        message: res.__("404")
+    })
+})
+
+export default router
```
Comment (Nkbtemmy) [inline src/routes/index.js]: remember to add new line

File: src/language/locales/en.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+	"passOemailInvalid": "Invalid Username or passoword",
+	"loginSuccess": "Login successful",
+	"Pleaselog": "Please log in or sign up to continue",
+	"PleaseVerify": "Please Verify your email to continue",
+	"log in to get access": "Please you may login to get access",
+	"permission for this action": "You do not have permission to perform this action",
+	"Token expired": "Your token expired, please login again",
+	"login again": "Please login again to continue",
+	"logout done": "logout done"
+}
```
Comment (Nkbtemmy) [inline src/language/locales/en.json]: remember to add new line

File: src/language/locales/fr.json [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+	"emptyPassORemail": "Remplissez les champs obligatoires",
+	"passOemailInvalid": "Nom d'utilisateur ou mot de passe invalide",
+	"loginSuccess": "connexion réussie",
+	"Pleaselog": "Veuillez vous connecter ou vous inscrire pour continuer",
+	"PleaseVerify": "Veuillez vérifier votre adresse e-mail pour continuer",
+	"log in to get access": "Veuillez vous connecter pour avoir accès",
+	"permission for this action": "Vous n'êtes pas autorisé à effectuer cette action",
+	"Token expired": "Votre token a expiré, veuillez vous reconnecter",
+	"login again": "Veuillez vous reconnecter pour continuer"
+}
```
Comment (Nkbtemmy) [inline src/language/locales/fr.json]: remember to add new line

File: src/test/testAuth.js [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+        let token = `eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MiwiaWF0IjoxNjAyNTk2OTU0LCJleHAiOjE2MTAzNzI5NTR9._Ygn8biPC6bpe4CYiD4sdH3fQR93u5hFaiWJXkpHheA`
+        chai.request(app)
+            .get('/api/v1/auth')
+            .set("Authorization", "Bearer " + token)
+            .end((err, res) => {
+                expect(res.body).to.be.an('object');
+                done();
+            });
+    });
+})
```
Comment (Nkbtemmy) [inline src/test/testAuth.js]: remember to add new line

File: src/middleware/protectRoutes.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        if (!roles.includes(req.user.role)) {
+            return next(
+                res.status(403).json({
+                    message: res.__('permission for this action')
+                })
+            )
+        }
+        next()
+    }
+}
```
Comment (Niyiragira) [inline src/middleware/protectRoutes.js:48]: I can see the new line issue is for many of the files

**Review by Nkbtemmy — APPROVED** [review]

**Review by karigirwa — APPROVED** [review]

**Review by Niyiragira — APPROVED** [review]

File: src/test/test.js [hunk 6]
```diff
@@ -51,7 +51,7 @@ describe('/POST Login user', () => {
      }
   it('it should check if  user exist', (done) => {
     chai.request(app)
-        .post('/api/v1/users/login')
+        .post('/api/v1/auth/login')
```
Comment (ceelogre) [inline src/test/test.js:54]: I want to think you spoke with the person who did signin before changing their code.

**Review by ceelogre — APPROVED** [review]

File: src/test/test.js (hunk 6 above)
Comment (ncutixavier) [reply src/test/test.js:54]: yeah, we did it together

**Pull request merged**